File: javascript-solutions/palindrome.js
const remove_non_alpha_numberic = (input_str) => {
  // remove_non_alpha_numberic is a helper function that remove all non alpha-numeric characters from a string
  const regex = /[^a-zA-Z0-9]/g;

  return input_str.replace(regex, "").toLowerCase();
};

const is_palindrome = (input_str) => {
  // A phrase is a palindrome if, after converting all uppercase letters into lowercase letters and removing all non-alphanumeric characters, it reads the same forward and backward.
  // Alphanumeric characters include letters and numbers.
  // Given a string s, return true if it is a palindrome, or false otherwise.

  // remove_non_alpha_numeric is a helper function that remove all non alpha-numeric characters from a string
  const processed_str = remove_non_alpha_numberic(input_str);

  let l_pointer = 0;
  let r_pointer = processed_str.length - 1;

  while (l_pointer !== r_pointer) {
    if (processed_str[l_pointer] !== processed_str[r_pointer]) return false;

    l_pointer += 1;
    r_pointer += 1;

    return true;
  }
};

console.log(is_palindrome("A man, a plan, a canal: Panama"));

File: python-solutions/palindrome.py
import re
def remove_non_alpha_numeric(input_str:str) -> str:
    """
        Returns a string that contains only alpha numeric characters
    """

    cleaned_sentence = re.sub(r'[^a-zA-Z0-9]', '', input_str)

    return cleaned_sentence.lower()


def is_palindrome(input_str:str) -> bool:
    
    '''
    A phrase is a palindrome if, after converting all uppercase letters into lowercase letters and removing all non-alphanumeric characters, it reads the same forward and backward. Alphanumeric characters include letters and numbers.

    Given a string s, return true if it is a palindrome, or false otherwise.

    >>> is_palindrome("A man, a plan, a canal: Panama")
    true

    >>> is_palindrome("race a car")
    false
    
    '''

    processed_str = remove_non_alpha_numeric(input_str)

    # print(processed_str)
    l,r = 0, len(processed_str) - 1 

    while l != r:
        if processed_str[l] != processed_str[r]:
            return False
        
        l += 1 
        r -= 1
        
    return True

  


if __name__ == "__main__":
    print(is_palindrome("race a car"))
    print(is_palindrome("A man, a plan, a canal: Panama"))
